CompositeCodec constructors require at least one delegate

cppwfs opened the pull request:
* In the current implementation, users could pass in an empty list to the composite codec.
* This is an CompositeCodec and if we had no codecs, then why have the class.
* All `Nullable's` were removed from the delegate attribute.
* Null returns from the findDelegate have been removed.   The default delegate will be returned in the stead.
* Update the tests to require at least one codec in the delegates

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

cppwfs committed:
Adjust the behavior of the class so that delegates are required.

* In the current implementation, users could pass in an empty list to the composite codec.
* This is an CompositeCodec and if we had no codecs, then why have the class.
* All @Nullable's were removed from the delegate attribute.
* Null returns from the findDelegate have been removed.   The default delegate will be returned in the stead.
* Update the tests to require at least one codec in the delegates

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -41,9 +39,20 @@ public class CompositeCodec implements Codec {
 	}
 
+	/**
+	 * @since 7.0
```
This tag means when the API was introduced.
Which is not correct in our case.
And doesn't matter no when it was introduced anyway since we are deprecating it.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -41,9 +39,20 @@ public class CompositeCodec implements Codec {
 
+	/**
+	 * @since 7.0
+	 *
```
No blank lines in the method JavaDoc.
I think our `src/idea/spring-framework.xml` does the trick to disable JavaDoc reformatting if that is the case.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -41,9 +39,20 @@ public class CompositeCodec implements Codec {
+	 * @deprecated This constructor is deprecated because the delegates map must contain at least one entry.
+	 * Creating a {@code CompositeCodec} without any delegates is not a valid use case, as there would be
+	 * no codecs to delegate to for encoding or decoding operations. Please use the constructor that accepts
+	 * a non-empty delegates map to ensure proper delegation behavior.
```
JavaDocs must be neutral and come with an official language without any `please` and similar.
How about this:
```
* @deprecated since 7.0 in favor of {@link #CompositeCodec(Map, Codec)} with some delegates.
```
I don't think we should say more here  about "why?".

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -41,9 +39,20 @@ public class CompositeCodec implements Codec {
+	 * a non-empty delegates map to ensure proper delegation behavior.
+	 *
+	 */
+	@Deprecated
```
Should be like: `@Deprecated(since = "7.0", forRemoval = true)`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -52,52 +61,35 @@ public CompositeCodec(Codec defaultCodec) {
-		}
-
+	private Codec findDelegate(Class<?> type) {
+		Assert.state(!this.delegates.isEmpty(), "Delegates must not be empty");
```
This is redundant. The `delegates` is initialized and validated in the ctor.
The state of the instance cannot be changed afterward.

artembilan reviewed (changes requested):
Please, add your name to the `@author` list of all the affected classes.

cppwfs committed:
Update code to meet code review requirements

cppwfs committed:
Add Author tag to affected classes

mrpiggi commented on spring-integration-core/src/test/java/org/springframework/integration/codec/kryo/CompositeCodecTests.java:
```diff
@@ -38,7 +38,8 @@ public class CompositeCodecTests {
 	public void setup() {
-		Map<Class<?>, Codec> codecs = new HashMap<>();
+		PojoCodec pojoCodec = new PojoCodec();
+		Map<Class<?>, Codec> codecs = Map.of(pojoCodec.getClass(), pojoCodec);
```
Adding to this: there is currently no test for the (now deprecated) ctor without a delegates map. Otherwise, the newly introduced assertion check ```java
Assert.notEmpty(delegates, "delegates must not be empty");
```
in the main ctor would have failed.

artembilan replied:
Correct.
We didn't have such a test and since we realized that ctor is wrong we are not going to add one.
But instead let's cover remaining one with more tests!

cppwfs committed:
Add tests to the CompositeCodecTests to tests delegates

And also test if a user attempts to an invalid class.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -52,52 +56,34 @@ public CompositeCodec(Codec defaultCodec) {
-		}
-
+	private Codec findDelegate(Class<?> type) {
 		Class<?> clazz = ClassUtils.findClosestMatch(type, this.delegates.keySet(), false);
```
I know, this is not the part of the problem we are fixing here, but I’d like to question this part. What does it mean and if it is explained in the JavaDocs and covered by tests.
Plus why there is no null check for this method call? Does it fail with error and that is what that your new code in the test? How does fallback to the default work then ?
Would you mind looking into that, please ?

mrpiggi replied:
I just saw, that I started a review on the former related PR but did not submit it so here my thoughts from the last review:

As `ClassUtils.findClosestMatch()` can return `null` (through `failOnTie=false`), one could argue that `Map::get` may not be called with `null` as key (which is no concern with `HashMap` but can be harmful if implementation is changed. So a safe approach would be
```java
	private Codec findDelegate(Class<?> type) {
		Class<?> clazz = ClassUtils.findClosestMatch(type, this.delegates.keySet(), false);
		return clazz == null ? this.defaultCodec : this.delegates.getOrDefault(clazz, this.defaultCodec);
	}
```

Regarding the question from Artem:

`ClassUtils.findClosestMatch()` traverses the class hierarchy by calling `ClassUtils.getTypeDifferenceWeight()` and gives back the class itself if found in `this.delegates.keySet()`. Otherwise, the "closest" super class or interface in the hierarchy found is returned.

As `failOnTie=false` is set, there will be something returned if a match in `this.delegates` was found. But as there could be multiple interfaces on the same "hierarchy level" in `this.delegates` and `ClassUtils.findClosestMatch()` is called with an unordered set, in this case the returned interface is non-deterministic. So one could argue that setting `failOnTie=true` would be the more defensive and reliable approach, though I have no suggestion, how one should then deal with the potentially thrown `IllegalStateException`

cppwfs replied:
Argue with me please.    We could give that option to the user  and they make the decision.   But we still support our original spec from the `Codec` interface:

```java
/**
* Finds the closest matching codec from the given set of candidate codecs based on the
* type difference weight between the specified target type and each candidate.
* Then encodes an object to a byte array using the chosen codec.
* @param object the object to encode
* @param failOnTie whether to fail with an exception if two codecs have equal minimal weight
* @return the bytes
* @throws IOException if the operation fails
*/  
  
public byte[] encode(Object object, boolean failOnTie) throws IOException {
	Assert.notNull(object, "cannot encode a null object");
	return findDelegate(object.getClass(), findOnTie).encode(object);
}

/**
* Finds the closest matching codec from the given set of candidate codecs based on the
* type difference weight between the specified target type and each candidate.
* Then encodes an object to a byte array using the chosen codec.  If 2 codes have the same weight, then one of the two will be chosen.
* @param object the object to encode
* @return the bytes
* @throws IOException if the operation fails
*/  
  
public byte[] encode(Object object) throws IOException {
	Assert.notNull(object, "cannot encode a null object");
	return findDelegate(object.getClass(), false).encode(object);
}
```

Thoughts?

cppwfs committed:
Update deprecated ctor so that it will only establish default codec

Per code review.   Thanks!

mrpiggi commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -41,63 +40,47 @@ public class CompositeCodec implements Codec {
 	public CompositeCodec(Codec defaultCodec) {
-		this(Map.of(), defaultCodec);
+		this.defaultCodec = defaultCodec;
+		this.delegates = Map.of();
```
That was my suggestion, but I had no look in the `6.5.x` branch so far. You are totally right, `new HashMap<>(null)` obviously cannot work. But then, this should be considered as a (10 year old) flawed/buggy implementation. And I still think, there are only two reasonable choices:

a) deprecate the ctor but provide a sane implementation
b) remove the flawed ctor completely without deprecation as this has never worked in the first place

I don't see, why an obvious bug should be kept although it was identified.

cppwfs committed:
Simplify the initialization of the `codec` attribute

cppwfs committed:
Update the javadoc for the CompositeCodec class.

Change the  `ClassUtils.findClosestMatch` `failOnTie` to true.
  In the past when finding the appropriate delegate(codec) for a object and
  their were multiple delegates available the first one was returned.
  The problem is it could return a different delegate on each launch of the `findDelegate`.
  Now with `failOnTie` set to true a `IllegalStateException` is returned.

Remove deprecated constructor from CompositeCodec

artembilan reviewed (changes requested):
And since we are here, please, consider to mention this class in the `codec.adoc`.
Thanks

cppwfs committed:
Replace sample classes in CompositeCodecTests with records to reduce maintenance

Rename field names of foo and foo2 in CompositeCodecTests to professional descriptive names
In the onlyDefaultCodec implementation in the CompositeCodecTests use a class other than a PojoCodec class to simulate a real world case.
Enhance javadoc for CompositeCodec to include information on ClassUtils.findClosestMatch and its effects on codec selection.
Make sure @see is in the correct location in javadoc

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/codec/CompositeCodec.java:
```diff
@@ -22,16 +22,28 @@
 import java.util.Map;
-
-import org.jspecify.annotations.Nullable;
+import java.util.Set;
```
I’m not sure how it works right now, but in the past it failed for us when is used only in JavaDocs. Or use FQCN in the link for the method, or mention that method without argument types.

artembilan reviewed (changes requested):
And add some paragraph with a sample code into `codec.adoc`, please.

cppwfs committed:
Add CompositeCodec documentation to reference documentation

Provide example for using CompositeCodec.
Remove @See in CompositeCodec javadocs.   This is handled in the links in the javadoc

cppwfs committed:
In codec.adoc all sentences must be on same line

Add code snippet to references to classes
Verify no tabs exist in sample

cppwfs commented on src/reference/antora/modules/ROOT/pages/codec.adoc:
````diff
@@ -53,6 +54,69 @@ The framework provides several custom serializers:
+An example implementation can be seen below:
+```java
+void encodeDecodeSample() {
+        Codec codec = getFullyQualifiedCodec();
````
No tabs were found in code segment.

cppwfs commented on src/reference/antora/modules/ROOT/pages/codec.adoc:
```diff
@@ -53,6 +54,69 @@ The framework provides several custom serializers:
 The second and third are used with the `MessageCodec`, which is initialized with the `MessageKryoRegistrar`.
 
+[[composite-codec]]
+=== CompositeCodec
```
Blank line applied

artembilan reviewed (approved):
LGTM.
But would be great to have https://github.com/spring-projects/spring-integration/pull/10207 reviewed and merged.
So, we could rebase this one before merge to working upstream.

Thanks
